rest/route/project.go

```diff
@@ -185,3 +186,133 @@ func (h *versionsGetHandler) Run(ctx context.Context) gimlet.Responder {
 
 	return gimlet.NewJSONResponse(versions)
 }
+
+/*
+ * Creates project (project_ref)
+ * PUT: /rest/v2/projects/
+ * Perm: superUser
+ */
+type projectCreateHandler struct {
+	projectRef *model.APIProjectRef
+
+	sc data.Connector
+}
+
+func makeProjectCreateRoute(sc data.Connector) gimlet.RouteHandler {
+	return &projectCreateHandler{sc: sc}
+}
+
+func (h *projectCreateHandler) Factory() gimlet.RouteHandler {
+	return &projectCreateHandler{sc: h.sc}
+}
+
+func (h *projectCreateHandler) Parse(ctx context.Context, r *http.Request) error {
+	body := util.NewRequestReader(r)
+	defer body.Close()
+
+	if err := util.ReadJSONInto(body, &h.projectRef); err != nil {
+		return errors.Wrap(err, "Cannot parse the content as JSON!")
+	}
+
+	return nil
+}
+
+func (h *projectCreateHandler) Run(ctx context.Context) gimlet.Responder {
+	createdProject, err := h.sc.CreateProject(h.projectRef)
+
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrap(err, "Cannot create project!"))
+	}
+
+	apiProject := &model.APIProject{}
+	// The method doesn't accept pointers, we have to dereference it first
+	err = apiProject.BuildFromService(*createdProject)
+
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			Message:    "problem converting project document",
+			StatusCode: http.StatusInternalServerError,
+		})
+	}
+
+	resp := gimlet.NewJSONResponse(apiProject)
+	err = resp.SetStatus(http.StatusCreated)
+
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			Message:    "Cannot set status code",
+			StatusCode: http.StatusInternalServerError,
+		})
+	}
+
+	return resp
+}
+
+/*
+ * Updates project (project_ref) with given identifier
+ * PATCH: /rest/v2/projects/{project_id}
+ * Perm: superUser
+ */
+type projectUpdateHandler struct {
+	projectRef *model.APIProjectRef
+
+	sc data.Connector
+}
+
+func makeProjectUpdateRoute(sc data.Connector) gimlet.RouteHandler {
+	return &projectUpdateHandler{sc: sc}
+}
+
+func (h *projectUpdateHandler) Factory() gimlet.RouteHandler {
+	return &projectUpdateHandler{sc: h.sc}
+}
+
+func (h *projectUpdateHandler) Parse(ctx context.Context, r *http.Request) error {
+	// Read raw bytes from request body
+	body := util.NewRequestReader(r)
+	defer body.Close()
+
+	projectRef := MustHaveProjectContext(ctx).ProjectRef
+	if projectRef == nil {
+		return errors.New("Cannot process request")
+	}
```

Comment:
I believe that MustHaveProjectContext panics if there's no project ref.

Reply:
It panics when there's no context value, it doesn't checks if tme `model.Context` has non-nil `*projectRef`.

Reply:
doesn't the middleware require that the project is populated? is there a way in which this isn't true?

Reply:
```
func MustHaveProjectContext(ctx context.Context) *model.Context {
	pc := GetProjectContext(ctx)
	if pc == nil {
		panic("project context not attached to request")
	}
	return pc
}
```
And
```
func GetProjectContext(ctx context.Context) *model.Context {
	p, _ := ctx.Value(RequestContext).(*model.Context)
	return p
}
```